src/script/tari_script.rs

```diff
@@ -138,41 +196,193 @@ impl TariScript {
+                true => stack.push(Number(1)),
+                false => stack.push(Number(0)),
             },
-            Opcode::CheckSigVerify => match self.check_sig(stack)? {
-                true => stack.push(Number(0)),
-                false => Err(ScriptError::InvalidSignature),
+            CheckSigVerify(msg) => match self.check_sig(stack, *msg.deref())? {
+                true => Ok(()),
+                false => Err(ScriptError::VerifyFailed),
             },
-            Opcode::RevRot => stack.push_down(2),
-            Opcode::PushHash(h) => stack.push(Hash(*h.clone())),
-            Opcode::HashBlake256 => TariScript::handle_hash::<Blake256>(stack),
-            Opcode::PushZero => stack.push(Number(0)),
+            Return => Err(ScriptError::Return),
+            IfThen => self.handle_if_then(stack),
+            Else => self.handle_else(),
+            EndIf => self.handle_end_if(),
+        }
+    }
+
+    fn handle_check_height_verify(height: u64, block_height: u64) -> Result<(), ScriptError> {
+        if block_height >= height {
+            Ok(())
+        } else {
+            Err(ScriptError::VerifyFailed)
+        }
+    }
+
+    fn handle_check_height(stack: &mut ExecutionStack, height: u64, block_height: u64) -> Result<(), ScriptError> {
+        let height = i64::try_from(height)?;
+        let block_height = i64::try_from(block_height)?;
+        let item = StackItem::Number(block_height - height);
+
+        stack.push(item)
+    }
+
+    fn handle_compare_height_verify(stack: &mut ExecutionStack, block_height: u64) -> Result<(), ScriptError> {
+        let item = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let target_height = if let StackItem::Number(n) = item {
+            Ok(n)
+        } else {
+            Err(ScriptError::InvalidInput)
+        }?;
+        let target_height = u64::try_from(target_height)?;
+
+        if block_height >= target_height {
+            Ok(())
+        } else {
+            Err(ScriptError::VerifyFailed)
+        }
+    }
+
+    fn handle_compare_height(stack: &mut ExecutionStack, block_height: u64) -> Result<(), ScriptError> {
+        let item = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let target_height = if let StackItem::Number(n) = item {
+            Ok(n)
+        } else {
+            Err(ScriptError::InvalidInput)
+        }?;
+        let block_height = i64::try_from(block_height)?;
+
+        let item = StackItem::Number(block_height - target_height);
+
+        stack.push(item)
+    }
+
+    fn handle_ordering(stack: &mut ExecutionStack, valid_orderings: Vec<Ordering>) -> Result<(), ScriptError> {
+        let item = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let stack_number = if let StackItem::Number(n) = item {
+            Ok(n)
+        } else {
+            Err(ScriptError::InvalidInput)
+        }?;
+
+        if valid_orderings.iter().any(|o| &stack_number.cmp(&0) == o) {
```

Comment:
Only need to do the `cmp` once

```suggestion
        let cmp = stack_number.cmp(0);
        if valid_orderings.contains(&cmp) {
```

Reply:
Nice I'll try that